drivers/pwm/pwm_ite_it51xxx.c

```diff
@@ -0,0 +1,313 @@
+#define REG_PWM_CTX_H(prs_sel)   (0x20 + 0x04 * (prs_sel) + 0x01)
+/* 0xF0: PWM global control */
+#define REG_PWM_GCTRL            0x70
+#define PWM_PCCE                 BIT(1)
+
+struct pwm_it51xxx_cfg {
+	/* PWM channel register base address */
+	uintptr_t base_ch;
+	/* PWM prescaler register base address */
+	uintptr_t base_prs;
+	/* Select PWM prescaler that output to PWM channel */
+	int prs_sel;
+	/* PWM alternate configuration */
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct pwm_it51xxx_data {
+	uint32_t ctx;
+	uint32_t pxc;
+	uint32_t target_freq_prev;
+};
+
+static void pwm_enable(const struct device *dev, int enabled)
+{
+	const struct pwm_it51xxx_cfg *config = dev->config;
+	const uintptr_t base_ch = config->base_ch;
+	uint8_t reg_val;
+
+	if (enabled) {
+		/* PWM channel clock source not gating */
+		reg_val = sys_read8(base_ch + REG_PWM_CH_CTRL0);
+		sys_write8(reg_val & ~PWM_CH_PCSG, base_ch + REG_PWM_CH_CTRL0);
+	} else {
+		/* PWM channel clock source gating */
+		reg_val = sys_read8(base_ch + REG_PWM_CH_CTRL0);
+		sys_write8(reg_val | PWM_CH_PCSG, base_ch + REG_PWM_CH_CTRL0);
+	}
+}
+
+static int pwm_it51xxx_get_cycles_per_sec(const struct device *dev, uint32_t channel,
+					  uint64_t *cycles)
+{
+	ARG_UNUSED(channel);
+
+	/*
+	 * Path of pwm_it51xxx_set_cycles() called from pwm api:
+	 * 1) pwm_set() -> pwm_set_cycles_cycles() -> pwm_it51xxx_set_cycles()
+	 *    target_freq = pwm_clk_src / period_cycles
+	 *                = cycles / (period * cycles / NSEC_PER_SEC)
+	 *                = NSEC_PER_SEC / period
+	 * 2) pwm_set_cycles() -> pwm_it51xxx_set_cycles()
+	 *    target_freq = pwm_clk_src / period_cycles
+	 *                = cycles / period
+	 *
+	 * When pwm needs output in EC power saving mode, we switch the prescaler
+	 * clock source (cycles) from 9.2MHz to 32.768kHz. Whether in power saving
+	 * mode or not, we need to get the same target_freq on above two cases,
+	 * so here always return PWM_FREQ.
+	 */
+	*cycles = (uint64_t)PWM_FREQ;
+
+	return 0;
+}
+
+static int pwm_it51xxx_set_cycles(const struct device *dev, uint32_t channel,
+				  uint32_t period_cycles, uint32_t pulse_cycles, pwm_flags_t flags)
+{
+	const struct pwm_it51xxx_cfg *config = dev->config;
+	const uintptr_t base_ch = config->base_ch;
+	const uintptr_t base_prs = config->base_prs;
+	struct pwm_it51xxx_data *data = dev->data;
+	int prs_sel = config->prs_sel;
+	uint32_t actual_freq = 0xffffffff, target_freq, deviation, dc_val;
+	uint64_t pwm_clk_src;
+	uint8_t reg_val;
+
+	/* Select PWM inverted polarity (ex. active-low pulse) */
+	if (flags & PWM_POLARITY_INVERTED) {
+		reg_val = sys_read8(base_ch + REG_PWM_CH_CTRL0);
+		sys_write8(reg_val | PWM_CH_INVP, base_ch + REG_PWM_CH_CTRL0);
+	} else {
+		reg_val = sys_read8(base_ch + REG_PWM_CH_CTRL0);
+		sys_write8(reg_val & ~PWM_CH_INVP, base_ch + REG_PWM_CH_CTRL0);
+	}
+
+	/* Enable PWM output open-drain */
+	if (flags & PWM_IT8XXX2_OPEN_DRAIN) {
```

Comment:
Please make it51xxx series have its own header file and define IT51XXX prefix macro in it.

Reply:
Done